src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -0,0 +1,279 @@
+        super(new ProductTypeSyncStatistics(), productTypeSyncOptions);
+        this.productTypeService = productTypeService;
+    }
+
+    /**
+     * Iterates through the whole {@code productTypeDrafts} list and accumulates its valid drafts to batches.
+     * Every batch is then processed by {@link ProductTypeSync#processBatch(List)}.
+     *
+     * <p><strong>Inherited doc:</strong>
+     * {@inheritDoc}
+     *
+     * @param productTypeDrafts {@link List} of {@link ProductTypeDraft}'s that would be synced into CTP project.
+     * @return {@link CompletionStage} with {@link ProductTypeSyncStatistics} holding statistics of all sync
+     *         processes performed by this sync instance.
+     */
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> process(
+            @Nonnull final List<ProductTypeDraft> productTypeDrafts) {
+
+        final List<List<ProductTypeDraft>> batches = batchElements(productTypeDrafts, syncOptions.getBatchSize());
+
+        return syncBatches(batches, CompletableFuture.completedFuture(statistics));
+    }
+
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> syncBatches(
+            @Nonnull final List<List<ProductTypeDraft>> batches,
+            @Nonnull final CompletionStage<ProductTypeSyncStatistics> result) {
+
+        if (batches.isEmpty()) {
+            return result;
+        }
+
+        final List<ProductTypeDraft> firstBatch = batches.remove(0);
+        return syncBatches(batches, result.thenCompose(subResult -> processBatch(firstBatch)));
+    }
+
+    /**
+     * Fetches existing {@link ProductType} objects from CTP project that correspond to passed {@code batch}.
+     * Having existing product types fetched, {@code batch} is compared and synced with fetched objects by
+     * {@link ProductTypeSync#syncBatch(List, List)} function. When fetching existing product types results in
+     * an empty optional then {@code batch} isn't processed.
+     *
+     * @param batch batch of drafts that need to be synced
+     * @return {@link CompletionStage} of {@link Void} that indicates method progress.
+     */
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> processBatch(@Nonnull final List<ProductTypeDraft> batch) {
+        final List<ProductTypeDraft> validProductTypeDrafts = batch.stream()
+                .filter(this::validateDraft)
+                .collect(toList());
+
+        if (validProductTypeDrafts.isEmpty()) {
+            statistics.incrementProcessed(batch.size());
+            return completedFuture(statistics);
+        } else {
+            final Set<String> keys = validProductTypeDrafts.stream().map(ProductTypeDraft::getKey).collect(toSet());
+
+            return fetchExistingProductTypes(keys)
+                .thenCompose(oldProductTypes -> syncBatch(oldProductTypes, validProductTypeDrafts))
+                .thenApply((ignored) -> {
```

Comment:
the parantheses around the `ignored` are not needed ;)

Reply:
cbc9b9699ecf191fe8a806e9df7c75ff8cc0afa8